src/libs/decoders/opus.c

```diff
@@ -0,0 +1,662 @@
+	                                     (size_t)nbytes);
+	SNDDBG(("Opus ops read:          "
+	        "{wanted: %d, returned: %ld}\n", nbytes, bytes_read));
+
+	return bytes_read;
+} /* RWops_opus_read */
+
+
+/*
+ * Read-Write Ops Seek Callback Wrapper
+ * ====================================
+ *
+ * OPUS: typedef int(* op_seek_func)
+ *       void*      _stream, --> The stream to seek in
+ *       opus_int64 _offset, --> Sets the position indicator for _stream in bytes
+ *       int _whence         --> If whence is set to SEEK_SET, SEEK_CUR, or SEEK_END,
+ *                               the offset is relative to the start of the stream,
+ *                               the current position indicator, or end-of-file,
+ *                               respectively
+ *    Returns: 0  Success, or -1 if seeking is not supported or an error occurred.
+ *      define	SEEK_SET	0
+ *      define	SEEK_CUR	1
+ *      define	SEEK_END	2
+ *
+ * SDL: Sint64 SDL_RWseek
+ *      SDL_RWops* context --> a pointer to an SDL_RWops structure
+ *      Sint64     offset, --> offset, in bytes
+ *      Sint32     whence  --> an offset in bytes, relative to whence location; can be negative
+ *   Returns the final offset in the data stream after the seek or -1 on error.
+ *      RW_SEEK_SET   0
+ *      RW_SEEK_CUR   1
+ *      RW_SEEK_END   2
+ */
+static Sint32 RWops_opus_seek(void* stream, const opus_int64 offset, const Sint32 whence)
+{
+	const Sint64 offset_after_seek = SDL_RWseek((SDL_RWops*)stream, (int)offset, whence);
+
+	SNDDBG(("Opus ops seek:          "
+	        "{requested offset: %ld, seeked offset: %ld}\n",
+	        offset, offset_after_seek));
+
+	return (offset_after_seek != -1 ? 0 : -1);
+} /* RWops_opus_seek */
+
+
+/*
+ * Read-Write Ops Close Callback Wrapper
+ * =====================================
+ * OPUS: typedef int(* op_close_func)(void *_stream)
+ * SDL:  Sint32 SDL_RWclose(struct SDL_RWops* context)
+ */
+static Sint32 RWops_opus_close(void* stream)
+{
+	/* SDL closes this for us */
+	// return SDL_RWclose((SDL_RWops*)stream);
+	return 0;
+} /* RWops_opus_close */
+
+
+/*
+ * Read-Write Ops Tell Callback Wrapper
+ * ====================================
+ * OPUS: typedef opus_int64(* op_tell_func)(void *_stream)
+ * SDL:  Sint64 SDL_RWtell(struct SDL_RWops* context)
+ */
+static opus_int64 RWops_opus_tell(void* stream)
+{
+	const Sint64 current_offset = SDL_RWtell((SDL_RWops*)stream);
+
+	SNDDBG(("Opus ops tell:          "
+	        "%ld\n", current_offset));
+
+	return current_offset;
+} /* RWops_opus_tell */
+
+
+// Populate the opus callback object (in perscribed order), with our callback functions.
+static const OpusFileCallbacks RWops_opus_callbacks =
+{
+	.read =  RWops_opus_read,
+	.seek =  RWops_opus_seek,
+	.tell =  RWops_opus_tell,
+	.close = RWops_opus_close
+};
+
+/* Return a human readable version of an OpusFile error code... */
+#if (defined DEBUG_CHATTER)
+static const char* opus_error(const Sint8 errnum)
```

Comment:
const char* opus_strerror(int error)

    Converts an opus error code into a human readable string. 

[link](https://opus-codec.org/docs/html_api-1.0.2/group__opus__libinfo.html#gafad3bac5a05dc7c3477a5765eb5e1873) Version of opus-devel (1.3.1) in Fedora does include this function, perhaps it was missing from some older version of the lib.

Reply:
Will swap it out; thanks!

Reply:
Fixed in https://github.com/dreamer/dosbox-staging/commit/2cb4a0a103a8d33d431f69a6483ff6dde682cb62